In [625]:
import kagglehub
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import xgboost as xgb
import gpboost as gpb
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score, RandomizedSearchCV
from sklearn.metrics import mean_squared_error

# Gathering the data

In [626]:
path = kagglehub.dataset_download("nishaanamin/march-madness-data")
print("Path to dataset files:", path)

Path to dataset files: /Users/kolbytaylor/.cache/kagglehub/datasets/nishaanamin/march-madness-data/versions/104


In [627]:
# Loading all the different data sets in

# Tournament Results and Data
coach_results = pd.read_csv("march-madness-data/versions/83/Coach Results.csv")
seed_results = pd.read_csv("march-madness-data/versions/83/Seed Results.csv")
team_results = pd.read_csv("march-madness-data/versions/83/Team Results.csv")
conference_results = pd.read_csv("march-madness-data/versions/83/Conference Results.csv")
tournament_locations = pd.read_csv("march-madness-data/versions/83/Tournament Locations.csv")
tournament_matchups = pd.read_csv("march-madness-data/versions/83/Tournament Matchups.csv")
upset_counts = pd.read_csv("march-madness-data/versions/83/Upset Count.csv")
upset_seed_info = pd.read_csv("march-madness-data/versions/83/Upset Seed Info.csv")

# Bartorvik
barttorvik_away_neutral = pd.read_csv("march-madness-data/versions/83/Barttorvik Away-Neutral.csv")
barttorvik_away = pd.read_csv("march-madness-data/versions/83/Barttorvik Away.csv")
barttorvik_home = pd.read_csv("march-madness-data/versions/83/Barttorvik Home.csv")
barttorvik_neutral = pd.read_csv("march-madness-data/versions/83/Barttorvik Home.csv")
kenpom_barttorvik = pd.read_csv("march-madness-data/versions/83/KenPom Barttorvik.csv")

# Conference Stats
conference_stats = pd.read_csv("march-madness-data/versions/83/Conference Stats.csv")
conference_away_neutral = pd.read_csv("march-madness-data/versions/83/Conference Stats Away Neutral.csv")
conference_away = pd.read_csv("march-madness-data/versions/83/Conference Stats Away.csv")
conference_home = pd.read_csv("march-madness-data/versions/83/Conference Stats Home.csv")
conference_neutral = pd.read_csv("march-madness-data/versions/83/Conference Stats Neutral.csv")

# Resumes
resumes = pd.read_csv("march-madness-data/versions/83/Resumes.csv")

# Shooting Splits
shooting_splits = pd.read_csv("march-madness-data/versions/83/Shooting Splits.csv")

# Other Websites
ratings_538 = pd.read_csv("march-madness-data/versions/83/538 Ratings.csv")
heat_check = pd.read_csv("march-madness-data/versions/83/Heat Check Tournament Index.csv")

# People's Opinions
preseason_votes = pd.read_csv("march-madness-data/versions/83/Preseason Votes.csv")
public_picks = pd.read_csv("march-madness-data/versions/83/Public Picks.csv")

In [628]:
# Take the columns in the dataframes that have home, away, and neutral; and subscript them so that we can recognize what it is
# Barttorvik
for col in barttorvik_away_neutral.columns:
    barttorvik_away_neutral = barttorvik_away_neutral.rename(columns={col: col + " AWAY-NEUTRAL"})

for col in barttorvik_away.columns:
    barttorvik_away = barttorvik_away.rename(columns={col: col + " AWAY"})

for col in barttorvik_home.columns:
    barttorvik_home = barttorvik_home.rename(columns={col: col + " HOME"})

for col in barttorvik_neutral.columns:
    barttorvik_neutral = barttorvik_neutral.rename(columns={col: col + " NEUTRAL"})



# Conference Statistics
for col in conference_away_neutral.columns:
    conference_away_neutral = conference_away_neutral.rename(columns={col: col + " AWAY-NEUTRAL"})

for col in conference_away.columns:
    conference_away = conference_away.rename(columns={col: col + " AWAY"})

for col in conference_home.columns:
    conference_home = conference_home.rename(columns={col: col + " HOME"})

for col in conference_neutral.columns:
    conference_neutral = conference_neutral.rename(columns={col: col + " NEUTRAL"})


In [629]:
barttorvik_away_neutral.head()

YEAR AWAY-NEUTRAL  TEAM NO AWAY-NEUTRAL  TEAM ID AWAY-NEUTRAL  \
0               2024                  1079                   197   
1               2024                  1078                    48   
2               2024                  1077                    33   
3               2024                  1076                    43   
4               2024                  1075                    36   

  TEAM AWAY-NEUTRAL  SEED AWAY-NEUTRAL  ROUND AWAY-NEUTRAL  \
0             Akron                 14                   0   
1           Alabama                  4                   0   
2           Arizona                  2                   0   
3            Auburn                  4                   0   
4            Baylor                  3                   0   

   BADJ EM AWAY-NEUTRAL  BADJ O AWAY-NEUTRAL  BADJ D AWAY-NEUTRAL  \
0                   3.1                105.0                101.9   
1                  18.7                125.4                106.7   
2                  23.7                119.7                 96.0   
3                  26.8                120.8                 94.0   
4                  17.6                117.7                100.1   

   BARTHAG AWAY-NEUTRAL  ...  BADJT RANK AWAY-NEUTRAL  \
0                 0.585  ...                      265   
1                 0.865  ...                       10   
2                 0.927  ...                       40   
3                 0.947  ...                       67   
4                 0.866  ...                      275   

   AVG HGT RANK AWAY-NEUTRAL  EFF HGT RANK AWAY-NEUTRAL  \
0                        238                        199   
1                         33                          8   
2                         50                         37   
3                         86                         76   
4                         31                         22   

   EXP RANK AWAY-NEUTRAL  TALENT RANK AWAY-NEUTRAL  FT% RANK AWAY-NEUTRAL  \
0                     19                       176                    186   
1                    156                       106                     12   
2                    196                         7                    157   
3                    127                        69                    127   
4                    304                        34                    137   

   OP FT% RANK AWAY-NEUTRAL  PPPO RANK AWAY-NEUTRAL  PPPD RANK AWAY-NEUTRAL  \
0                       188                     120                      47   
1                       357                       3                     351   
2                       284                      19                      98   
3                       260                       5                      18   
4                       248                      52                     189   

   ELITE SOS RANK AWAY-NEUTRAL  
0                          269  
1                            8  
2                           52  
3                           77  
4                           11  

[5 rows x 85 columns]

There are many columns in the Barttorvik data, and splitting them out by where the team played multiplies it by five. In total we have around 170 features. Half of them are metrics and the other half rank those metrics in all Division 1. We should probably sort these out first. I'm going to try doing some analysis with the features. I'm going to fit the same model with the overall, away, netural, home, and neutral-away features and see which ones work the best, as well as do some AB testing to see if there's a significant difference between them.

In [630]:
# Creates the target, 0-6 based on how many games were won
conditions = [
    (kenpom_barttorvik['ROUND'] == 0) | (kenpom_barttorvik['ROUND'] == 68),
    (kenpom_barttorvik['ROUND'] == 64),
    (kenpom_barttorvik['ROUND'] == 32),
    (kenpom_barttorvik['ROUND'] == 16),
    (kenpom_barttorvik['ROUND'] == 8),
    (kenpom_barttorvik['ROUND'] == 4),
    (kenpom_barttorvik['ROUND'] == 2),
    (kenpom_barttorvik['ROUND'] == 1)
]

values = [np.nan, 0, 1, 2, 3, 4, 5, 6]

kenpom_barttorvik['TARGET'] = np.select(conditions, values, default=np.nan)
barttorvik_away_neutral['TARGET'] = np.select(conditions, values, default=np.nan)
barttorvik_away['TARGET'] = np.select(conditions, values, default=np.nan)
barttorvik_home['TARGET'] = np.select(conditions, values, default=np.nan)
barttorvik_neutral['TARGET'] = np.select(conditions, values, default=np.nan)

In [631]:
barttorvik_away_neutral.head()

YEAR AWAY-NEUTRAL  TEAM NO AWAY-NEUTRAL  TEAM ID AWAY-NEUTRAL  \
0               2024                  1079                   197   
1               2024                  1078                    48   
2               2024                  1077                    33   
3               2024                  1076                    43   
4               2024                  1075                    36   

  TEAM AWAY-NEUTRAL  SEED AWAY-NEUTRAL  ROUND AWAY-NEUTRAL  \
0             Akron                 14                   0   
1           Alabama                  4                   0   
2           Arizona                  2                   0   
3            Auburn                  4                   0   
4            Baylor                  3                   0   

   BADJ EM AWAY-NEUTRAL  BADJ O AWAY-NEUTRAL  BADJ D AWAY-NEUTRAL  \
0                   3.1                105.0                101.9   
1                  18.7                125.4                106.7   
2                  23.7                119.7                 96.0   
3                  26.8                120.8                 94.0   
4                  17.6                117.7                100.1   

   BARTHAG AWAY-NEUTRAL  ...  AVG HGT RANK AWAY-NEUTRAL  \
0                 0.585  ...                        238   
1                 0.865  ...                         33   
2                 0.927  ...                         50   
3                 0.947  ...                         86   
4                 0.866  ...                         31   

   EFF HGT RANK AWAY-NEUTRAL  EXP RANK AWAY-NEUTRAL  TALENT RANK AWAY-NEUTRAL  \
0                        199                     19                       176   
1                          8                    156                       106   
2                         37                    196                         7   
3                         76                    127                        69   
4                         22                    304                        34   

   FT% RANK AWAY-NEUTRAL  OP FT% RANK AWAY-NEUTRAL  PPPO RANK AWAY-NEUTRAL  \
0                    186                       188                     120   
1                     12                       357                       3   
2                    157                       284                      19   
3                    127                       260                       5   
4                    137                       248                      52   

   PPPD RANK AWAY-NEUTRAL  ELITE SOS RANK AWAY-NEUTRAL  TARGET  
0                      47                          269     NaN  
1                     351                            8     NaN  
2                      98                           52     NaN  
3                      18                           77     NaN  
4                     189                           11     NaN  

[5 rows x 86 columns]

In [632]:
X_barttorvik = kenpom_barttorvik.drop(columns=['YEAR', 'CONF', 'CONF ID', 'QUAD NO', 'QUAD ID', 'TEAM NO', 'TEAM ID', 'TEAM', 'SEED', 'ROUND', 'TARGET'])
y_barttorvik = kenpom_barttorvik['TARGET']

X_barttorvik_away_neutral = barttorvik_away_neutral.drop(columns=['YEAR AWAY-NEUTRAL', 'TEAM NO AWAY-NEUTRAL', 'TEAM ID AWAY-NEUTRAL', 'TEAM AWAY-NEUTRAL', 'SEED AWAY-NEUTRAL', 'ROUND AWAY-NEUTRAL', 'TARGET'])
y_barttorvik_away_neutral = barttorvik_away_neutral['TARGET']

X_barttorvik_away = barttorvik_away.drop(columns=['YEAR AWAY', 'TEAM NO AWAY', 'TEAM ID AWAY', 'TEAM AWAY', 'SEED AWAY', 'ROUND AWAY', 'TARGET'])
y_barttorvik_away = barttorvik_away['TARGET']

X_barttorvik_home = barttorvik_home.drop(columns=['YEAR HOME', 'TEAM NO HOME', 'TEAM ID HOME', 'TEAM HOME', 'SEED HOME', 'ROUND HOME', 'TARGET'])
y_barttorvik_home = barttorvik_home['TARGET']

X_barttorvik_neutral = barttorvik_neutral.drop(columns=['YEAR NEUTRAL', 'TEAM NO NEUTRAL', 'TEAM ID NEUTRAL', 'TEAM NEUTRAL', 'SEED NEUTRAL', 'ROUND NEUTRAL', 'TARGET'])
y_barttorvik_neutral = barttorvik_neutral['TARGET']

# Drop rows where TARGET is null
X_barttorvik = X_barttorvik.loc[y_barttorvik.notna()]
y_barttorvik = y_barttorvik.dropna()

X_barttorvik_away_neutral = X_barttorvik_away_neutral.loc[y_barttorvik_away_neutral.notna()]
y_barttorvik_away_neutral = y_barttorvik_away_neutral.dropna()

X_barttorvik_away = X_barttorvik_away.loc[y_barttorvik_away.notna()]
y_barttorvik_away = y_barttorvik_away.dropna()

X_barttorvik_home = X_barttorvik_home.loc[y_barttorvik_home.notna()]
y_barttorvik_home = y_barttorvik_home.dropna()

X_barttorvik_neutral = X_barttorvik_neutral.loc[y_barttorvik_neutral.notna()]
y_barttorvik_neutral = y_barttorvik_neutral.dropna()


I don't want to have to do feature scaling or do anything with assumptions because this is just EDA modeling, so I'm going to use XGBoost.

In [633]:
# No null values! That's nice. 
print(X_barttorvik.isnull().sum()[X_barttorvik.isnull().sum() > 0])

Series([], dtype: int64)


In [634]:
# Datatypes look good
print(X_barttorvik.dtypes)

K TEMPO           float64
K TEMPO RANK        int64
KADJ T            float64
KADJ T RANK         int64
K OFF             float64
                   ...   
FT% RANK            int64
OP FT% RANK         int64
PPPO RANK           int64
PPPD RANK           int64
ELITE SOS RANK      int64
Length: 93, dtype: object


In [635]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik, y_barttorvik, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24614
[10]	eval-rmse:1.07732
[20]	eval-rmse:1.08317
[30]	eval-rmse:1.08900
[33]	eval-rmse:1.09478
Best iteration: 14
Test RMSE: 1.0766


In [636]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik_away_neutral, y_barttorvik_away_neutral, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24819
[10]	eval-rmse:1.08775
[20]	eval-rmse:1.08954


[30]	eval-rmse:1.09811
[33]	eval-rmse:1.10664
Best iteration: 13
Test RMSE: 1.0897


In [637]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik_away, y_barttorvik_away, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24884


[10]	eval-rmse:1.12063
[20]	eval-rmse:1.11433
[30]	eval-rmse:1.12300
[33]	eval-rmse:1.13158
Best iteration: 14
Test RMSE: 1.1112


In [638]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik_home, y_barttorvik_home, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24086
[10]	eval-rmse:1.07804
[20]	eval-rmse:1.08300
[30]	eval-rmse:1.09856
Best iteration: 10
Test RMSE: 1.0840


In [639]:
# Train-test split (80-20)
X_train, X_test, y_train, y_test = train_test_split(X_barttorvik_neutral, y_barttorvik_neutral, test_size=0.2, random_state=42)

# Prepare DMatrix (required for xgb.train)
dtrain = xgb.DMatrix(X_train, label=y_train)
dtest = xgb.DMatrix(X_test, label=y_test)

# Define hyperparameters grid manually
params = {
    'objective': 'reg:squarederror',
    'eval_metric': 'rmse',
    'max_depth': 2,
    'learning_rate': 0.2,
    'subsample': 0.6,
    'alpha': 10,
    'colsample_bytree': 0.6,
    'random_state': 42,
    'min_child_weight': 5
}

# Set up training with early stopping
evallist = [(dtest, 'eval')]

# Train the model with early stopping
bst = xgb.train(
    params, 
    dtrain, 
    num_boost_round=1000,
    evals=evallist, 
    early_stopping_rounds=20,
    verbose_eval=10
)

# Get the best iteration from early stopping
best_iteration = bst.best_iteration
print(f"Best iteration: {best_iteration}")

# Make predictions using the best iteration
y_pred = bst.predict(dtest, iteration_range=(0, best_iteration))

# Calculate RMSE on test set
test_rmse = np.sqrt(mean_squared_error(y_test, y_pred))

# Print the result
print(f"Test RMSE: {test_rmse:.4f}")


[0]	eval-rmse:1.24086
[10]	eval-rmse:1.07804
[20]	eval-rmse:1.08300
[30]	eval-rmse:1.09856
Best iteration: 10
Test RMSE: 1.0840


### Model Results:

The results were consistent across all models. The hyperparameters selected by each model were identical:

- subsample: 0.6  
- min_child_weight: 5  
- max_depth: 2  
- learning_rate: 0.05  
- lambda: 5  
- colsample_bytree: 1.0  
- alpha: 10  

---

### RMSE:

- Overall: 1.07  
- Away-Neutral: 1.08  
- Away: 1.11  
- Home: 1.08  
- Neutral: 1.08  

---

The Overall RMSE might be slightly better due to the inclusion of a few different features. Based on this analysis, we have decided to use the overall Barttorvik data in our analysis moving forward.


In [640]:
# Starting with the Kenpom data
all_data = kenpom_barttorvik

# Add 538 Ratings
all_data = all_data.merge(ratings_538, how='left', on='TEAM NO')
all_data = all_data.drop(columns=['YEAR_y', 'TEAM_y', 'SEED_y', 'ROUND_y'])
all_data = all_data.rename(columns={'YEAR_x': 'YEAR', 'TEAM_x': 'TEAM', 'SEED_x': 'SEED', 'ROUND_x': 'ROUND'})

# Add Heat Check metrics
all_data = all_data.merge(heat_check, how='left', on='TEAM NO')
all_data = all_data.drop(columns=['YEAR_y', 'TEAM_y', 'SEED_y', 'ROUND_y'])
all_data = all_data.rename(columns={'YEAR_x': 'YEAR', 'TEAM_x': 'TEAM', 'SEED_x': 'SEED', 'ROUND_x': 'ROUND'})

# Add resume data, includes strength of schedule
all_data = all_data.merge(resumes, how='left', on='TEAM NO')
all_data = all_data.drop(columns=['YEAR_y', 'TEAM_y'])
all_data = all_data.rename(columns={'YEAR_x': 'YEAR', 'TEAM_x': 'TEAM', 'SEED_x': 'SEED', 'ROUND_x': 'ROUND'})

# Shooting splits of each team
all_data = all_data.merge(shooting_splits, how='left', on='TEAM NO')
all_data = all_data.drop(columns=['YEAR_y', 'TEAM_y', 'SEED_y', 'ROUND_y'])
all_data = all_data.rename(columns={'YEAR_x': 'YEAR', 'TEAM_x': 'TEAM', 'SEED_x': 'SEED', 'ROUND_x': 'ROUND', 'CONF_x': 'CONF', 'TEAM ID_x': 'TEAM ID'})

# Drops any more columns that don't seem useful
all_data = all_data.drop(columns=['CONF ID', 'QUAD NO', 'TEAM ID', 'TARGET', 'VAL Z-SCORE', 'ROUND', 'TEAM ID_y', 'CONF_y'])

all_data['new_seeding_code'] = all_data['SEED'].astype(str) + all_data['QUAD ID'].astype(str)

mapping = {
    '14': 1,   '11': 2,   '12': 3,   '13': 4,
    '23': 5,   '22': 6,   '21': 7,   '24': 8,
    '34': 9,   '31': 10,  '32': 11,  '33': 12,
    '43': 13,  '42': 14,  '41': 15,  '44': 16,
    '54': 17,  '51': 18,  '52': 19,  '53': 20,
    '63': 21,  '62': 22,  '61': 23,  '64': 24,
    '74': 25,  '71': 26,  '72': 27,  '73': 28,
    '83': 29,  '82': 30,  '81': 31,  '84': 32,
    '94': 33,  '91': 34,  '92': 35,  '93': 36,
    '103': 37, '102': 38, '101': 39, '104': 40,
    '114': 41, '111': 42, '112': 43, '113': 44,
    '123': 45, '122': 46, '121': 47, '124': 48,
    '134': 49, '131': 50, '132': 51, '133': 52,
    '143': 53, '142': 54, '141': 55, '144': 56,
    '154': 57, '151': 58, '152': 59, '153': 60,
    '163': 61, '162': 62, '161': 63, '164': 64
}

all_data['SEED'] = all_data['new_seeding_code'].map(mapping)

all_data = all_data.drop(columns=['new_seeding_code'])


There are coach and conference results that seem cool and would be interesting, but scare me because of potential data leakage. For example, Bill Self won a championship a few years ago, but before that his championships would be at zero... So I don't think it's a good idea to include it. Conference results are similar, but might be more watered down because they're average over the entire conference.

In [641]:
# Prints the number of columns
print(len(all_data.columns))

153


Our feature space is really high, so we're going to try and reduce it by removing the columns that rank the features. These are obviously highly correlated with the features themselves, we're just losing some information about how the team compared to other teams in its season.

In [642]:
all_data = all_data.drop(columns=[col for col in all_data.columns if 'RANK' in col])

In [643]:
# Prints the number of columns
print(len(all_data.columns))

89


Okay, now we're down to a more reasonable space. I'm sure there's some unimportant features, or features with high collinearity that we can weed out, but we'll have to more to find them.

In [644]:
all_data.to_csv("All_Data.csv", index=False)

In [645]:
all_data

YEAR  CONF  QUAD ID  TEAM NO              TEAM  SEED  K TEMPO   KADJ T  \
0     2024   MAC        1     1079             Akron  55.0  66.7747  65.8933   
1     2024   SEC        3     1078           Alabama  13.0  74.1625  72.6461   
2     2024   P12        3     1077           Arizona   5.0  73.3760  71.8379   
3     2024   SEC        4     1076            Auburn  16.0  70.9629  69.7887   
4     2024   B12        3     1075            Baylor  12.0  66.8428  65.6032   
...    ...   ...      ...      ...               ...   ...      ...      ...   
1074  2008    BE        1        5     West Virginia  26.0  66.2223  63.0942   
1075  2008    SB        1        4  Western Kentucky  47.0  68.7887  66.5432   
1076  2008  BSth        4        3          Winthrop  49.0  63.7666  60.9746   
1077  2008   B10        3        2         Wisconsin  12.0  62.1920  60.8809   
1078  2008   A10        1        1            Xavier  10.0  65.1599  62.8978   

        K OFF    KADJ O  ...  CLOSE TWOS FG%D  CLOSE TWOS D SHARE  \
0     107.841  107.0090  ...             58.0                37.7   
1     121.712  125.6010  ...             60.2                37.9   
2     117.653  121.1250  ...             57.4                30.3   
3     117.364  120.5790  ...             48.3                42.0   
4     117.262  122.4900  ...             65.7                29.8   
...       ...       ...  ...              ...                 ...   
1074  111.246  113.5700  ...              NaN                 NaN   
1075  110.914  111.0760  ...              NaN                 NaN   
1076  101.646   99.6932  ...              NaN                 NaN   
1077  108.495  113.6190  ...              NaN                 NaN   
1078  115.214  116.4130  ...              NaN                 NaN   

      FARTHER TWOS FG%  FARTHER TWOS SHARE  FARTHER TWOS FG%D  \
0                 43.8                24.6               39.9   
1                 35.7                10.8               37.6   
2                 37.9                23.1               38.5   
3                 41.6                21.1               33.6   
4                 37.5                26.9               40.0   
...                ...                 ...                ...   
1074               NaN                 NaN                NaN   
1075               NaN                 NaN                NaN   
1076               NaN                 NaN                NaN   
1077               NaN                 NaN                NaN   
1078               NaN                 NaN                NaN   

      FARTHER TWOS D SHARE  THREES FG%  THREES SHARE  THREES FG%D  \
0                     25.6        32.0          41.5         30.0   
1                     25.3        36.5          46.8         31.9   
2                     31.5        37.1          32.6         33.4   
3                     24.7        35.2          37.5         29.8   
4                     34.6        39.4          38.3         33.5   
...                    ...         ...           ...          ...   
1074                   NaN         NaN           NaN          NaN   
1075                   NaN         NaN           NaN          NaN   
1076                   NaN         NaN           NaN          NaN   
1077                   NaN         NaN           NaN          NaN   
1078                   NaN         NaN           NaN          NaN   

      THREES D SHARE  
0               36.8  
1               36.7  
2               38.2  
3               33.3  
4               35.5  
...              ...  
1074             NaN  
1075             NaN  
1076             NaN  
1077             NaN  
1078             NaN  

[1079 rows x 89 columns]

In [646]:
quadrants = kenpom_barttorvik[['TEAM NO', 'QUAD ID']]

tournament_matchups = tournament_matchups.merge(
    quadrants, how='left', on='TEAM NO', suffixes=('', '_drop')
).drop(columns=['QUAD ID_drop'], errors='ignore')

In [647]:
ids_to_remove = [2034, 2033, 2028, 2027, 1980, 1979, 1976, 1975]
tournament_matchups = tournament_matchups[~tournament_matchups['BY YEAR NO'].isin(ids_to_remove)]

In [648]:
odd_index_matchups = tournament_matchups.iloc[::2]
even_index_matchups = tournament_matchups.iloc[1::2]

odd_index_matchups = odd_index_matchups.reset_index(drop=True)
odd_index_matchups = odd_index_matchups.merge(all_data, how='left', on='TEAM NO')

even_index_matchups = even_index_matchups.reset_index(drop=True)
even_index_matchups = even_index_matchups.merge(all_data, how='left', on='TEAM NO')

odd_index_matchups.columns = [col + '_1' for col in odd_index_matchups.columns]
even_index_matchups.columns = [col + '_2' for col in even_index_matchups.columns]

combined_matchups = pd.concat([odd_index_matchups, even_index_matchups], axis=1)
columns_to_drop = [
    'YEAR_y_1', 
    'YEAR_x_2', 
    'YEAR_y_2', 
    'TEAM_y_1', 
    'TEAM_y_2', 
    'BY YEAR NO_1', 
    'BY YEAR NO_2', 
    'BY ROUND NO_1', 
    'BY ROUND NO_2', 
    'TEAM NO_1', 
    'TEAM NO_2',
    'SEED_x_1',
    'SEED_x_2',
    'QUAD ID_x_1',
    'QUAD ID_x_2',
    'QUAD ID_y_1',
    'QUAD ID_y_2',
    'ROUND_1',
    'ROUND_2',
    'CURRENT ROUND_2',
    'BID TYPE_1',
    'BID TYPE_2'
]
existing_columns_to_drop = [col for col in columns_to_drop if col in combined_matchups.columns]
combined_matchups = combined_matchups.drop(columns=existing_columns_to_drop)

combined_matchups = combined_matchups.rename(columns={
    'YEAR_x_1': 'YEAR',
    'TEAM_x_1': 'TEAM_1',
    'TEAM_x_2': 'TEAM_2',
    'SEED_y_1': 'SEED_1',
    'SEED_y_2': 'SEED_2',
    'CURRENT ROUND_1': 'ROUND'
})

desired_order = ['YEAR', 'ROUND', 'CONF_1', 'TEAM_1', 'SEED_1', 'SEED_2', 'CONF_2', 'TEAM_2']
remaining_columns = [col for col in combined_matchups.columns if col not in desired_order]
combined_matchups = combined_matchups[desired_order + remaining_columns]

columns_to_process = [col for col in remaining_columns if '_1' in col]

for col_1 in columns_to_process:
    col_2 = col_1.replace('_1', '_2')
    if col_2 in combined_matchups.columns:
        combined_matchups[col_1.replace('_1', '') + '_diff'] = (combined_matchups[col_1].astype(float) - combined_matchups[col_2].astype(float)).round(3)
        combined_matchups = combined_matchups.drop(columns=[col_1, col_2])

# Filter out 2008 and 2009 since shooting splits aren't available
combined_matchups = combined_matchups[combined_matchups['YEAR'] >= 2010]

combined_matchups['target'] = np.where(combined_matchups['SCORE_diff'].isnull(), np.nan, 
                                       np.where(combined_matchups['SCORE_diff'] > 0, 1, 0))

combined_matchups

YEAR  ROUND CONF_1          TEAM_1  SEED_1  SEED_2 CONF_2  \
0    2024     64     BE     Connecticut     1.0    64.0   ASun   
1    2024     64    ACC  North Carolina     4.0    61.0    NEC   
2    2024     64    B12         Houston     3.0    62.0   BSth   
3    2024     64    B10          Purdue     2.0    63.0   SWAC   
4    2024     64    B12        Iowa St.     8.0    57.0    Sum   
..    ...    ...    ...             ...     ...     ...    ...   
883  2010      8    SEC        Kentucky     3.0     6.0     BE   
884  2010      8    ACC            Duke     2.0    10.0    B12   
885  2010      4    B10    Michigan St.    17.0    20.0   Horz   
886  2010      4     BE   West Virginia     6.0     2.0    ACC   
887  2010      2   Horz          Butler    20.0     2.0    ACC   

               TEAM_2  SCORE_diff  K TEMPO_diff  ...  CLOSE TWOS D SHARE_diff  \
0             Stetson         NaN        -0.397  ...                     -0.4   
1              Wagner         NaN         8.892  ...                     -3.7   
2            Longwood         NaN        -3.734  ...                      1.2   
3       Grambling St.         NaN         2.695  ...                     -6.6   
4    South Dakota St.         NaN         0.036  ...                     -3.9   
..                ...         ...           ...  ...                      ...   
883     West Virginia        -7.0         6.214  ...                     -3.8   
884            Baylor         7.0         0.265  ...                     12.3   
885            Butler        -2.0         2.311  ...                     -0.5   
886              Duke       -21.0        -3.704  ...                      1.0   
887              Duke        -2.0        -3.375  ...                     -3.0   

     FARTHER TWOS FG%_diff  FARTHER TWOS SHARE_diff  FARTHER TWOS FG%D_diff  \
0                     -7.9                     -6.9                   -10.2   
1                      7.7                     -4.9                    -4.6   
2                     -3.6                      5.0                    -8.6   
3                      1.4                     -2.3                    -3.0   
4                     -7.4                     13.3                    -1.4   
..                     ...                      ...                     ...   
883                    0.1                     -3.7                    -2.6   
884                   -6.6                     -0.9                    -3.9   
885                    3.5                     13.1                     4.4   
886                    3.5                     -2.1                     2.3   
887                    5.0                    -13.8                     0.7   

     FARTHER TWOS D SHARE_diff  THREES FG%_diff  THREES SHARE_diff  \
0                          4.6              0.2               -0.9   
1                          8.8              3.8               -2.3   
2                         -2.7              0.3                8.1   
3                          2.8              6.7                5.2   
4                          6.0             -1.3               -8.2   
..                         ...              ...                ...   
883                        1.6             -0.5               -3.0   
884                       -3.2              0.3                0.6   
885                       -4.7              0.2              -13.6   
886                       -7.9             -4.6                1.7   
887                       -3.5             -4.2                6.9   

     THREES FG%D_diff  THREES D SHARE_diff  target  
0                -2.6                 -4.2     NaN  
1                 1.5                 -5.0     NaN  
2                -4.5                  1.6     NaN  
3                -2.3                  3.8     NaN  
4                -3.7                 -2.1     NaN  
..                ...                  ...     ...  
883              -1.0                  2.2     0.0  
884              -5.4    

In [649]:
combined_matchups.to_csv("combined_matchups.csv", index=False)

# EDA